OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/ide/instances/lcp_on_fields/problem/LCPOnFieldsProblem.scala

```diff
@@ -0,0 +1,831 @@
+import org.opalj.br.IntegerType
+import org.opalj.br.Method
+import org.opalj.br.analyses.DeclaredFieldsKey
+import org.opalj.br.analyses.SomeProject
+import org.opalj.br.fpcf.properties.immutability.EffectivelyNonAssignable
+import org.opalj.br.fpcf.properties.immutability.FieldAssignability
+import org.opalj.br.fpcf.properties.immutability.LazilyInitialized
+import org.opalj.br.fpcf.properties.immutability.NonAssignable
+import org.opalj.fpcf.FinalP
+import org.opalj.fpcf.InterimUBP
+import org.opalj.fpcf.PropertyStore
+import org.opalj.ide.problem.EdgeFunctionResult
+import org.opalj.ide.problem.FinalEdgeFunction
+import org.opalj.ide.problem.FlowFunction
+import org.opalj.ide.problem.InterimEdgeFunction
+import org.opalj.ide.problem.MeetLattice
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.LinearConstantPropagationPropertyMetaInformation
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.LinearConstantPropagationLattice
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.LinearConstantPropagationValue
+import org.opalj.tac.fpcf.analyses.ide.problem.JavaIDEProblem
+import org.opalj.tac.fpcf.analyses.ide.solver.JavaICFG
+import org.opalj.tac.fpcf.analyses.ide.solver.JavaStatement
+import org.opalj.tac.fpcf.analyses.ide.solver.JavaStatement.StmtAsCall
+
+/**
+ * Definition of linear constant propagation on fields problem. This implementation detects basic cases of linear
+ * constant propagation involving fields. It detects direct field assignments but fails to detect assignments done in a
+ * method where the value is passed as an argument (e.g. a classical setter). Similar, array read accesses can only be
+ * resolved if the index is a constant literal. There also is just minimal support for static fields.
+ * This implementation is mainly intended to be an example of a cyclic IDE analysis.
+ *
+ * @author Robin Körkemeier
+ */
+class LCPOnFieldsProblem(
+    project: SomeProject,
+    icfg:    JavaICFG
+) extends JavaIDEProblem[LCPOnFieldsFact, LCPOnFieldsValue] {
+    private val declaredFields = project.get(DeclaredFieldsKey)
+
+    override val nullFact: LCPOnFieldsFact =
+        NullFact
+
+    override val lattice: MeetLattice[LCPOnFieldsValue] =
+        LCPOnFieldsLattice
+
+    override def getAdditionalSeeds(stmt: JavaStatement, callee: Method)(
+        implicit propertyStore: PropertyStore
+    ): collection.Set[LCPOnFieldsFact] = {
+        (if (callee.isStatic) {
+             immutable.Set.empty
+         } else {
+             /* Add fact for `this` */
+             immutable.Set(ObjectFact("param0", -1))
+         }) ++
+            /* Add facts for other parameters */
+            callee.parameterTypes
+                .zipWithIndex
+                .filter { case (paramType, _) => paramType.isObjectType || paramType.isArrayType }
```

Comment:
Consider replacing filter+map with collect, plus performing the whole operation on an iterator